packages/@react-stately/table/src/useColumnResizeWidthState.ts

```diff
@@ -0,0 +1,227 @@
+
+import {ColumnProps} from '@react-types/table';
+import {GridNode} from '@react-types/grid';
+import {Key, MutableRefObject, useEffect, useRef, useState} from 'react';
+
+export default function useColumnResizeWidthState<T>(
+    columns: GridNode<T>[],
+    getDefaultWidth: (props) => string | number
+  ): [MutableRefObject<Map<Key, number>>, (column: GridNode<T>, newWidth: number) => { key: Key, width: number }[], (width: number) => void] {
+
+  const tableWidth = useRef<number>(null);
+  const [columnWidths, setColumnWidths] = useState<Map<Key, number>>(initializeColumnWidths(columns));
+  const columnWidthsRef = useRef<Map<Key, number>>(columnWidths);
+  const [resizedColumns, setResizedColumns] = useState<Set<Key>>(new Set());
+  const resizedColumnsRef = useRef<Set<Key>>(resizedColumns);
+
+  // if the columns change, need to rebuild widths.
+  useEffect(() => {
+    const widths = buildColumnWidths(columns, tableWidth.current);
+    setColumnWidthsForRef(widths);
+  }, [columns]);
+  
+
+  function initializeColumnWidths(columns: GridNode<T>[]): Map<Key, number> {
+    const widths = new Map();
+    columns.forEach(column => widths.set(column.key, 0));
+    return widths;
+  }
+
+  function setColumnWidthsForRef(newWidths: Map<Key, number>) {
+    columnWidthsRef.current = newWidths;
+    // new map so that change detection is triggered
+    setColumnWidths(newWidths);
+  }
+
+  function setTableWidth(width: number) {
+    if (width && width !== tableWidth.current) {
+      tableWidth.current = width;
+      const widths = buildColumnWidths(columns, width);
+      setColumnWidthsForRef(widths);
+    }
+  }
+
+  function resizeColumn(column: GridNode<T>, newWidth: number) : { key: Key, width: number }[] {
+    // copy the columnWidths map and set the new width for the column being resized
+    let widths = new Map<Key, number>(columnWidthsRef.current);
+    widths.set(column.key, Math.max(
+      getMinWidth(column.props.minWidth),
+      Math.min(roundWidth(newWidth), getMaxWidth(column.props.maxWidth))
+    ));
+
+    // keep track of all columns that have been seized
```

Comment:
I think double clicking to fit content (like excel) would be awesome. It sounds like a really hard problem though because we have to find out the exact size that the largest of the columns' cells content wants to be. We assumed there's no easy way to do that and be performant but maybe you know of a good way to do it? I think it would be a great nice-to-have.